walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,85 +14,114 @@
+      }
 
-        // Update the latest conflict end time seen so far.
-        int currentConflictEnd = currentConflict.end();
-        if (currentConflictEnd > lastConflictEnd) {
-            lastConflictEnd = currentConflictEnd;
-        }
+      int currentConflictEnd = currentConflict.end();
+      if (currentConflictEnd > lastConflictEnd) {
+        lastConflictEnd = currentConflictEnd;
+      }
     }
 
     // Check if the time between the latest end time and the end of the day is a valid time range.
     if (TimeRange.END_OF_DAY - lastConflictEnd >= requestDuration) {
-        timeRanges.add(TimeRange.fromStartEnd(lastConflictEnd, TimeRange.END_OF_DAY, /* inclusive= */ true));
+      timeRanges.add(
+          TimeRange.fromStartEnd(lastConflictEnd, TimeRange.END_OF_DAY, /* inclusive= */ true));
     }
     return timeRanges;
   }
 
-  private static List<TimeRange> getPotentialConflicts(Collection<Event> events, 
-        Collection<String> requestAttendees) {
+  private static List<TimeRange> getPotentialConflicts(
+      Collection<Event> events, Collection<String> requestAttendees) {
     List<TimeRange> potentialConflicts = new ArrayList<TimeRange>();
     for (Event event : events) {
-        // Find the intersection between current event attendees and request attendees.
-        Set<String> intersection = new HashSet<String>(event.getAttendees());
-        intersection.retainAll(requestAttendees);
+      // Find the intersection between current event attendees and request attendees.
+      Set<String> intersection = new HashSet<String>(event.getAttendees());
+      intersection.retainAll(requestAttendees);
 
-        // If the intersection is not empty, add this event time to the potential
-        // conflicts.
-        if (!intersection.isEmpty()) {
-            potentialConflicts.add(event.getWhen());
-        }
-    } 
+      // If the intersection is not empty, add this event time to the potential
+      // conflicts.
+      if (!intersection.isEmpty()) {
+        potentialConflicts.add(event.getWhen());
+      }
+    }
     return potentialConflicts;
   }
+  /*
+   * Return the time ranges where all mandatory attendees can attend and the maximum
+   * number of optional attendees can attend.
+   */
+  private static Collection<TimeRange> getTimeRangesWithMaximumOptionalAttendees(
+      Collection<Event> events, Collection<TimeRange> mandatoryTimeRanges,
+      Collection<String> optionalAttendees) {
+    List<TimeRange> result = new ArrayList<TimeRange>();
+    int minOptionalAttendeesWithConflict = Integer.MAX_VALUE;
+    for (TimeRange timeRange : mandatoryTimeRanges) {
+      Set<String> optionalAttendeesWithConflict = new HashSet<String>();
+      for (Event e : events) {
+        if (timeRange.contains(e.getWhen())) {
+          for (String attendee : e.getAttendees()) {
+            if (optionalAttendees.contains(attendee)) {
+              optionalAttendeesWithConflict.add(attendee);
+            }
+          }
```

Comment:
This can be shorten by stream
e.g.
e.getAttendees().stream().filter(attendee -> optionalAttendees.contains(attendee)).collect(toHashSet).

Copied from best practice
When should I use streams?
For many operations, streams can be extremely compact, efficient, and easy to use. Certainly streams are not appropriate for all operations, and if you encounter difficulty migrating a certain block of code to use streams, it may be the case that streams just aren't the right tool for that job. If an operation can be rewritten as a simple, compact stream expression, however, it's not a bad idea to do so.

For code that already uses FluentIterable or transform and filter-like methods from com.google.common.collect, streams are very likely superior. These existing methods generally return views, where streams can only return concrete collections or results, but many of these cases can still be rewritten to streams. For cases where it is clearly semantics-preserving, we have provided tools to rewrite that code to streams automatically.

Reply:
Done - I see, thank you.